### PR TITLE
Improve GraphQL typing

### LINKED ISSUE #55: Improve GraphQL operations typing
I discussed this with @rdig and understand there were some issues with implementing GraphQL code generator in the same way it was implemented in the dapp, ie. generating custom hooks for each operation, e.g. `useColonyFromNameQuery`, `useUserAddressQuery`, etc.
As a result, the data returned from the `useQuery` is of type `any`.
One of the proposed solutions was to manually create wrapper hooks for the queries that would also provide typings for the response.

### PR DESCRIPTION
**Pull request by jakubcolony**
## Description

This PR attempts to improve the GraphQL typing by replacing the Amplify's codegen with the `graphql-codegen` package which was also used in the dapp.

You can run it using `npm run codegen`. 

I also added it to the Amplify Docker image so it runs automatically on any changes to `*.graphql` files. ~~There's one caveat with this though, it doesn't pick up the Amplify's schema the first time it runs but I don't think it will cause a lot of headache since there's always gonna be the some version of `generated.ts` file present, and if you make any changes subsequently, this will trigger another codegen run.~~ You might see a few codegen errors while it tries to fetch the schema but as soon as Amplify GraphQL API is fully up, the codegen will run.

~~I wasn't able to fully disable the Amplify's codegen. You will see the following error message in the console:
It doesn't seem to impact anything else however.~~
Thanks to Raul for his suggestion I was able to silence those by creating a fake schema and config files.

### Motivation

To understand some of my changes, I think it's useful to first get a grasp of how things used to work with the amplify codegen:
1. You add your models to `schema.graphql`. Amplify generates some basic GraphQL CRUD operations.
2. Amplify's codegen generates not only typings for object/input types but also actual queries and mutations. However, by default, they select all the fields of a particular type, therefore making it much less useful as you'd typically want to select just a subset of fields.
3. If you wanted to amend the generated query in any way (e.g. omitting some fields, renaming them, etc.), you had to add it to `custom-queries.ts` file.

This is what a fully typed query would look like:
```
import { gql, useQuery } from '@apollo/client';

import {
  getFullColonyByName,
  GetFullColonyByNameQuery,
  GetFullColonyByNameQueryVariables,
} from '~gql';

// inside component
const { data } = useQuery<
  GetFullColonyByNameQuery,
  GetFullColonyByNameQueryVariables
>(gql(getFullColonyByName), {
  variables: {
    name: '...',
  },
});
```

### New solution

By using `graphql-codegen` we can achieve the following workflow: 
1. Same as above.
2. You define the exact query you want to make in `~gql/queries/*.graphql`.
3. `graphql-codegen` generates the typings and hooks corresponding to the operations you defined.

You end up with fully-typed reusable hooks:
```
import { useGetFullColonyByName } from '~gql';

// inside component
const {} = useGetFullColonyByName({
  variables: {
    name: '...',
  },
});
```

Additionally, by using fragments you get the correct types for the subset of fields you're querying. 

### Changes to how Amplify docker image's package.json is generated
Following Raul's suggestions, I implemented the following workflow: 
 - Amplify image dependencies should be specified in `src/docker/files/amplify/dependencies` (one per line)
 - As part of the build process, `generate-package-json` script is ran that creates the `package.json` file by merging `package.base.json` with the list of dependencies and their versions from CDapp's `package.json`

## TODO

- [x] Remove Amplify codegen
- [x] Run codegen from within Docker container automatically 

Resolves #55

### CODE REVIEW

File: src/components/common/ColonyHome/ColonyFundingWidget/ColonyFundingWidget.tsx
Comment (jakubcolony) [inline src/components/common/ColonyHome/ColonyFundingWidget/ColonyFundingWidget.tsx:38]: @rdig I'm not a fan of how Amplify resolves many-to-many relations. You'd expect the `tokens` field to contain `Token` object, but it's actual return type is `ModelColonyTokensConnection`. Maybe we should rename it at the query level?

File: .graphqlconfig.yml
Comment (rdig) [inline .graphqlconfig.yml:1]: I know how to get rid of your "providers" error.

Screenshot from 2022-11-22 01-38-23

- create a `fakeschema.graphql` file which needs to be empty
- restore this file
- replace `schemaPath` with the path to the empty fake schema

Error gone and amplify's internal codegen will do nothing since the file is empty. :tada: 

For even more bonus points, in an effort to keep everything clean, create both this file, and the fake schema just inside the doker container, so that the codebase won't get polluted by useless files, and yet amplify's requirements will be met.

**Review by rdig — CHANGES_REQUESTED** [review]
Mostly good with this, and I really like the improvements this brings, but I'm not allowing this to be merged until we iron out the kinks, namely:
- the codegen race condition
- the providers amplify error

File: docker/files/amplify/package.json.base
Comment (jakubcolony) [reply docker/files/amplify/package.json.base]: I think if I went down the `jq`/`sed` route we'd end up with a lot of boilerplate code (repeated for every dependency) so I took a slightly different approach:
- You define the dependencies in a simple textfile called `dependencies`
- Instead of running `jq`/`sed` and writing `package.json` directly, Docker image runs `generate-package-json` script that goes through every dependency and injects version from CDapp's `package.json`
- It merges the resolved dependencies with `package.base.json` which can still be used to define scripts and anything else

Let me know your thoughts...

File: docker/files/amplify/run.sh.base
Comment (jakubcolony) [reply docker/files/amplify/run.sh.base]: That was way more trickier than I thought.

Amplify is pretty quick to start sending HTTP 200s and it does it even before it creates the DB tables which results in codegen still failing. I tried to detect when the `/graphql` endpoint is ready (which itself wasn't straightforward, the endpoint returns 404 for GET requests and 401 for POST without the auth header) but that would still happen before the tables were created. 

At this point I think I ran out of sensible ideas on how to detect when Amplify is fully finished with its setup and the API is ready to be queried. For now, I added a 10s delay after waiting for Amplify with `wait-on` which does the trick, but it will likely fail on slower machines or when we add new models.

File: src/types/domains.ts
Comment (jakubcolony) [reply src/types/domains.ts:57]: >Maybe an IIFE that generates the object at run time ?

But we'd still need the mapping somewhere, right?

The only improvement I can think of is to start using the GraphQL's `DomainColor` enum instead of a custom one, that way we'll only need to maintain the mapping between `DomainColor` and hex colors, instead of DomainColor <> Color <> hex mappings

File: docker/files/amplify/run.sh.base
Comment (jakubcolony) [reply docker/files/amplify/run.sh.base]: Actually, I might be able to write a custom codegen script that retries fetching the schema when it fails: https://the-guild.dev/graphql/codegen/docs/advanced/programmatic-usage. 

I'll look into that tomorrow.

**Review by willm30 — APPROVED** [review]
Nice improvement! Use of auto-generated hooks is very convenient, as is use of fragments / ability to increase query specificity. Tested by adding a `GetFullColonyByAddress` query to `colony.graphql` and the hooks generation worked like a charm. 

There will be a few more components that can be refactored off the back of this issue, e.g. the Colony Creation wizard which has been committed to master since this branch was put up for review.  How do you want to handle updating those? In a separate issue? I'm happy to create.

File: src/components/common/ColonyHome/ColonyFundingWidget/ColonyFundingWidget.tsx
Comment (chinins) [reply src/components/common/ColonyHome/ColonyFundingWidget/ColonyFundingWidget.tsx:44]: @rdig, so should we all follow this approach for filtering? It is applicable to all queries that return list and I think it's good to agree on one approach?

**Review by chinins — APPROVED** [review]
Great work - I'm looking forward to using it in my PRs :) will wait on rebasing my current PR until you merge this one so that I can implement the changes.

File: docker/colony-cdapp-dev-env-amplify
Comment (rdig) [inline docker/colony-cdapp-dev-env-amplify:23]: Please keep the source file naming similar and add `.base` to the end of their name/extension

File: docker/colony-cdapp-dev-env-amplify
Comment (rdig) [inline docker/colony-cdapp-dev-env-amplify:19]:
```suggestion
ADD docker/files/amplify/package.json.base /colonyCDapp/package.json.base
```

File: src/components/common/ColonyHome/ColonyFundingWidget/ColonyFundingWidget.tsx
Comment (rdig) [reply src/components/common/ColonyHome/ColonyFundingWidget/ColonyFundingWidget.tsx:44]: Yes, if you might have `null` items returned

**Review by rdig — APPROVED** [review]
Right, I'm good with this now.

Please address the fine naming things I pointed out below, after which you can rebase/merge it.

PS: after rebase, please make sure the current existing queries / mutations that the CDapp uses are all still working, and if not please go in and fix them.

**Pull request merged**
